ClinGen/clincoded
# Connects to #1328. Allow adding diseases by other ontologies and free text in addition to Orphanet.

**Technical notes:**
This PR introduces a huge amount of new and updated code that are fundamentally changing how the GCI/VCI allows curators add/associate diseases given the curation context. The test steps provided below may not be able to cover every change made in the process, or all possible test cases.

_Some of the notable changes are:_
1. Deprecating the use of locally stored Orphanet Phenotypes and using the MonDO merge ontology REST API at OLS to look up disease terms.
2. In addition to allowing 4 ontologies to look up disease terms (e.g. Orphanet, DOID, OMIM and NCIt) via the MonDO REST API, curators can now add free text diseases.
3. The various interfaces affected by these changes are GDM, Group, Family, Individual, CaseControl, and Interpretation (of VCI).

**Steps to test GDM:**
1. When creating a new GDM, expect to see a **Disease+** button replacing the _Orphanet ID_ text field.
2. Click on the **Disease+** button and expect to see the "Add Disease" being displayed.
3. Enter an ontology ID per example and click on the **Retrieve from OLS** button. Expect to see the disease info being returned and displayed in the modal.
4. Click the **Save** button in the modal and expect to see the returned disease info being rendered on the Create-GDM page.
5. Click on the **Disease[edit]** button and expect to see the modal again in which you may edit by looking up a different ontology ID.
6. Click on the **Reset** button in the modal and expect to see the pre-filled form being cleared and now you may choose to enter free text disease by selecting the checkbox in the lower half of the modal.
7. Fill out the required fields for the free text disease and click the **Save** button in the modal. Expect to see the free text disease info being rendered on the Create-GDM page.
8. Proceed with filling out other required fields and expect to be able save the Create-GDM page.

**Steps to test a created GDM:**
1. After successfully saving the newly created GDM, while on the Curation-Central page and prior to adding any PMIDs to this GDM, expect to see the **Disease[edit]** button in the header next to the disease term.
2. Click on the **Disease[edit]** button and expect to see the modal in which you may edit the previously saved free text disease or change to an ontology-based disease term via the **Reset** button.
3. After the edit, click the **Save** button and expect to see the changed disease info being rendered in the header as well as in the DiseaseRecord header slightly below.
4. If the saved disease term has a definition, expect to see the [View definition] link being displayed in the DiseaseRecord header. Click on this link and expect to see a popover showing the disease definition.
5. Now add a PMID to this GDM and expect to see the **Disease[edit]** button no longer being shown in the header.

**Steps to test Group:**
1. In a new Group evidence form, expect to see the **Disease+** button for entering disease. Also expect to see the **Copy disease from Gene-Disease record** button that allows you to copy the disease saved for the GDM into this Group evidence.
2. After a disease being copied or added, expect to see the **Delete** button next to it. Click on the **Delete** button and expect to see the disease being removed.
3. Fill out the other _required_ fields **without** a disease. Save the form and expect to see error messages indicating that a disease is required for the Group evidence.
4. Now copy or add a disease. Expect to be able to successfully save the form and proceed to adding a Family evidence to this Group evidence.

**Steps to test Family:**
1. In the associated Family evidence form, expect to see the **Disease+** button for entering disease. Also expect to see the **Copy disease from associated Group** button that allows you to copy the disease saved for the Group into this Family evidence.
2. After a disease being copied or added, expect to see the **Delete** button next to it. Click on the **Delete** button and expect to see the disease being removed. **Keep in mind that a disease is not _required_ for a Family evidence.**
3. Again copy or add a disease to the Family, and fill out other _required_ fields.
4. Copy or add a disease for the Proband without entering a name. Save the form and expect to see error messages indicating that a Proband name is required (if a disease is added for the Proband).
5. Enter the Proband name and delete the Proband disease (via the **Delete** button). Save the form and expect to see error messages indicating that the Proband disease is required (if a Proband name is entered for the Family).
6. Now add the Proband disease and expect to be able to save the Family evidence form successfully.

**Steps to test Individual:**
1. In the associated Individual evidence form, expect to see that the disease has been pre-populated with the previously saved Proband disease from the Family evidence.
2. Click on the **Delete** button and expect to see the disease being removed, as well as the **Copy disease from associated Family** and **Disease+** buttons being shown.
3. In a new Individual evidence without any associated Group or Family, and expect to see the **Copy disease from Gene-Disease record** button being shown.
4. Add a new Individual evidence to the created Group, and expect to see the **Copy disease from associated Group** button being shown on the evidence form.

**Steps to test CaseControl:**
1. In a new CaseControl form, expect to see the **Copy disease from Gene-Disease record** and **Disease+** buttons for entering disease in Case-Cohort.
2. After a disease being copied or added in Case-Cohort, expect to see the **Delete** button below the disease info. Click on the **Delete** button and expect to see the disease being removed.
3. Fill out the other _required_ fields **without** a disease. Save the form and expect to see error messages indicating that a disease is required for the Case-Cohort.
4. Now copy or add a disease, and expect to be able to successfully save the form.

**Steps to test Interpretation:**
1. In VCI, add a variant and start a new interpretation. Expect to see the **Disease+** button being shown at the upper right.
2. Click on the **Disease+** button and expect to see the modal being shown for adding either an ontology-based disease term or free text disease.
3. Fill out the _required_ fields and click the **Save** button in the modal. Expect to see the newly associated disease info being rendered in the header, as well as the **Delete** button now being shown at the upper right.
